### PR TITLE
[CDU] Add Type I and Type II Messages

### PR DESCRIPTION
**Pull request by MisterChocker**
## Summary of Changes

- Type I Messages gets deleted on input to mcdus scratchpad.
- Type II Messages disappear when Type I Messages are displayed.
- Type II Messages disappear on user input and become visible again when scratchpad is empty again.
- Type II Messages are cleared by CLR key.
- Type II Messages are displayed first-in/first-out.
- Queue can hold 5 Type II Messages, if more are added, oldest are deleted from queue.
- Updated GPS Primary Lost to use Type II functionality.
- Old message handler now handles Type I Messages.
- ~~Port other messages over to new system.~~ (Will do in additional pr).

Old error message handling is still valid and used for TYPE I message handling.

| usage | mcdu function name | parameter a | parameter b | parameter c |
| ----- | ----- | ----- | ----- | ----- |
| Type I | `showErrorMessage` | message | color _(optional)_| none |
| Type II | `addTypeIIMessage` | message | color _(optional)_ | function which gets executed on message clear _(optional)_ |

## References

image

## Additional context

Discord username (if different from GitHub): DerL30N#3751

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js [hunk 1]
```diff
@@ -300,15 +307,24 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
...
+                    console.log("primary");
+                    this.tryRemoveMessage("GPS PRIMARY LOST");
+                    this.showErrorMessage("GPS PRIMARY");
+                }
             } else {
-                this.showErrorMessage("GPS PRIMARY");
+                if (!SimVar.GetSimVarValue("L:A32NX_GPS_PRIMARY_LOST_MSG", "Bool")) {
+                    SimVar.SetSimVarValue("L:A32NX_GPS_PRIMARY_LOST_MSG", "Bool", 1);
+                    SimVar.SetSimVarValue("L:GPSPrimaryMessageDisplayed", "Bool", 0);
+                    console.log("primary lost");
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:322]: Remove `console.log` calls

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js [hunk 2]
```diff
@@ -320,6 +336,69 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
...
+        this.isDisplayingErrorMessage = false;
+    }
+
+    /**
+     * Add Type II Message
+     * @param message {string} Message to be displayed
+     * @param color {string} Color of Message
+     * @param f {function} Function gets executed when error message has been cleared
+     */
+    addTypeIIMessage(message, color = "#ffffff", f = () => {}) {
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:352]: Can we change this to `addTypeTwoMessage` ?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 3]
```diff
@@ -3110,6 +3113,7 @@ class FMCMainDisplay extends BaseAirliners {
             if (this.onInputAircraftSpecific(input)) {
                 return;
             }
+            console.log("input: " + input);
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:3116]: Remove `console.log` calls

**Review by Benjozork — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js (hunk 2 above)
Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js]: yep

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by beheh** [comment]
Related, but possibly out of scope for this PR: in the MCDU docs, some of the type II messages have either EFIS PFD or EFIS ND set. The EFIS ND ones are the ones that will appear on the ND (like GPS PRIMARY), and the PFD ones are the ones that appear at the top of the PFD (like DECELERATE).

That could be an extra parameter here to `addTypeTwoMessage`, for example:

```
    addTypeTwoMessage(message, color = "#ffffff", f = () => {}, efisMode = 0) {
```

where

```
MCDUMessageEFISMode.NONE = 0
MCDUMessageEFISMode.EFIS_PFD = 1
MCDUMessageEFISMode.EFIS_ND = 2
```

**Comment by MisterChocker** [comment]
@beheh I am sorry but I don't understand what you try to explain. We are already displaying messages on ND.

**Comment by beheh** [comment]
So right now the messages are hardoded in the MFD code: 

https://github.com/flybywiresim/a32nx/blob/87d1583bd595ecee9e76e9403c114361c6bd741f/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/MFD/A320_Neo_MFD.html#L136-L137

But actually it's just a property of a message whether it should appear on the PFD, on the ND, or not at all. Messages that have it, show it in the Honeywell guide:

grafik

These messages are closely linked to the MCDU, as they disappear from the ND (and presumably the ND) when they're cleared from the MCDU. So my suggestion was to get started on the API in place to optionally mark messages for the PFD/ND, and then we can use your dismiss type II message logic instead of manually using and clearing separate simvars for every single message (like `GPSPrimaryAcknowledged`).

Although I am slightly confused why `GPS PRIMARY LOST` and `GPS PRIMARY` do not have (EFIS ND) each in the documentation.

In any case, I think that's work for a future PR.

**Comment by MisterChocker** [comment]
yeah definitely out of scope of this pr. But good a got issue for another pr!

**Comment by Benjozork** [comment]
For QA, only check general functionality of MCDU and error messages on bad format, etc.

**Comment by HugoTro** [comment]
Name: Hugo #3307
Date of testing: Current date 17/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1951 

Testing technique: Observation and engine fire
Steps you've made to test the Pull Request:
1. Spawned in approach of LFRS RWY 03 with engine 1 fire armed
2. Actually, the engine started burning just after loading
3. looked at the ECAM and observed that i could clear the fire message/instructions (which i think it's type I) but the "land asap" still persisted
4. Landed
5. Brakes hot in amber showed up but i was still able to clear them (which i think it's type I)

Overall Rating: Good?
Conclusions of the testing: Let me know if fire instructions/brakes hot messages are supposed to be deleted just by pressing CLR

**Comment by MisterChocker** [comment]
@HugoTro This is mcdu not ecam. @Benjozork gave instructions on how to test. (One comment above yours)

**Comment by HugoTro** [comment]
where do i need to look at in the MCDU?

**Comment by HugoTro** [comment]
Name: Hugo #3307
Date of testing: Current date 17/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1951 

Testing technique: Look at
Steps you've made to test the Pull Request:
1. C&D at LFMN Gate 37
2. Looked at options to make an amber message appear (they were all not clearable by CLR) 

Overall Rating: To be improved
Conclusions of the testing: The only amber message that could be deleted was V1/VR/V2 DISAGREE, all the white messages were deleted with CLR

**Comment by MisterChocker** [comment]
@HugoTro 
> 2. Looked at options to make an amber message appear (they were all not clearable by CLR)

Which messages could not be cleared?

**Comment by Carter633** [comment]
Name: British Avgeek UTube -- Carter633 #1436
Date of testing: Current date 18/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1951 
Object of test: Type 1 & 2 Messages into MCDU
Testing technique: Short Full Flight EGLL - EGPH
Steps you've made to test the Pull Request:
1. Spawned cold and dark, EGLL - worked through MCDU and intentionally tried incorrect inputs to check for issues
2. Completed MCDU input for full flight correctly, depart
3. Tried to trigger a queue of messages to push the MCDU - no issues
4. Checked for any MCDU changes elsewhere
5. Landed

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: Seems fine, though I don't know the technical requirements of it I went through the MCDU and all worked fine, even when trying to put in duff invalid lines and trying to muck with the messages. No issues found!

**Comment by MisterChocker** [comment]
@Carter633 sounds great! Wondering what wasn't working for @HugoTro

**Comment by Carter633** [comment]
I think he was meant to type that he was not able to clear the V speed Disagree message but it came out wrong

**Comment by CaptainR3X** [comment]
Category: QA Tester
Name: GustavRex#4347
Date of testing: Current date 18/11/2020
Version of the sim: 1.10.11.0
PR Tested: #1951
Object of test: Type I and Type II Messages in MCDU

Testing technique: Flight; EKCH-ESGG
Steps you've made to test the Pull Request:
1. C/D at EKCH, stress tested MCDU to provoke a faliure by inserting wrong values and reprogramming flightplan incorrectly several time
2. MCDU flight programming, p/s, taxi and departure
3. Initiated several wrong procedures to provoke overload leading to faliures
4. Arrival programming, following same procedure as in the beginning
5. Landing, taxi and shut-down

Flight notes (If required): Online, VatSim - multiple DIR and reprogramming 

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: During the testing I tried my hardest to provoke any kind of issues with the MCDU. MSG requests, wrong loads and values and inserting wrong waypoints to further mess with the computer. I was unable to find any faliures on the MCDU on any part of the flight.

**Comment by MisterChocker** [comment]
@Benjozork Any reason this is marked as tested? Are we waiting on any decisions?

**Comment by Benjozork** [comment]
I forgot to merge it

**Pull request merged**